kroxylicious/src/main/java/io/kroxylicious/proxy/internal/DefaultFilterContext.java

```diff
@@ -104,13 +173,16 @@ public void forwardResponse(ApiMessage response) {
         }
         // TODO check we've not forwarded it already
 
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug("{}: Forwarding response: {}", channelDescriptor(), decodedFrame);
+        }
         channelContext.fireChannelRead(decodedFrame);
     }
 
     private void checkNotClosed() {
-        if (closed) {
-            throw new IllegalStateException("Context is closed");
-        }
+        // if (closed) {
+        // throw new IllegalStateException("Context is closed");
+        // }
```

Comment:
I really hated doing this. The problem is that when the filter calls `sendRequest` it returns and the `context` gets closed, but the lambda used for the completion of the `Future` captures the `context` and needs to be used to call `forwardResponse` once we've looked processed the response (e.g. looked up the topic name).

So
* either we need to provide a fetch context to the handler (i.e. `sendRequest()` returns a `Future<Something>`, where something includes the response data and the new context.
* or we use reference counting, so that `sendRequest()` increments the `count` and `close()` decrements it and the check becomes one for `>0`. 

I think I prefer the latter option. The problem with the first is that it would still be a common programming error to accidentally use the captured context.

Reply:
Hum, hum. I'm a bit unclear on the scope/lifectime of filter contexts anyways. Can't they be scoped to the connection?

Reply:
Yeah, I think I'm coming to the same conclusion. There's nothing actually wrong with using a `KrpcFilterContext` after an `on*Request()` or `on*Response()` method has returned, at least while the connection/channel still exists. The problem I'm really wanting to solve is how to make programming using this filter API as simple as possible, with sensible guardrails for simple programming mistakes. 

Using `forwardRequest` or `forwardResponse` multiple times would be one thing to avoid, but it comes at the cost of needing some additional state (e.g. on the Frame), so I've not done that yet.